orchagent/sfloworch.cpp

```diff
@@ -0,0 +1,347 @@
+            {
+                rate = (uint32_t)stoul(fvValue(i));
+            }
+            else
+            {
+                rate = 0;
+            }
+        }
+    }
+}
+
+void SflowOrch::sflowStatusSet(Consumer &consumer)
+{
+    auto it = consumer.m_toSync.begin();
+
+    while (it != consumer.m_toSync.end())
+    {
+        auto tuple = it->second;
+        string op = kfvOp(tuple);
+        uint32_t rate = 0;
+
+        if (op == SET_COMMAND)
+        {
+            sflowExtractFvs (kfvFieldsValues(tuple), sflowStatus, rate);
+        }
+        else if (op == DEL_COMMAND)
+        {
+            sflowStatus = false;
+        }
+        it = consumer.m_toSync.erase(it);
+    }
+}
+
+uint32_t SflowOrch::sflowSessionGetRate(sai_object_id_t m_sample_id)
+{
+    for (auto it: m_sflowRateSampleMap)
+    {
+        if (it.second.m_sample_id == m_sample_id)
+        {
+            return it.first;
+        }
+    }
+    return 0;
+}
+
+bool SflowOrch::handleSflowSessionDel(sai_object_id_t port_id)
+{
+    auto sflowInfo = m_sflowPortInfoMap.find(port_id);
+
+    if (sflowInfo != m_sflowPortInfoMap.end())
+    {
+        uint32_t rate = sflowSessionGetRate(sflowInfo->second.m_sample_id);
+        if (sflowInfo->second.admin_state)
+        {
+            if (!sflowDelPort(port_id))
+            {
+                return false;
+            }
+            sflowInfo->second.admin_state = false;
+        }
+
+        m_sflowPortInfoMap.erase(port_id);
+        m_sflowRateSampleMap[rate].ref_count--;
+        if (m_sflowRateSampleMap[rate].ref_count == 0)
+        {
+            if (!sflowDestroySession(m_sflowRateSampleMap[rate]))
+            {
+                return false;
+            }
+            m_sflowRateSampleMap.erase(rate);
+        }
+    }
+    return true;
+}
+
+void SflowOrch::doTask(Consumer &consumer)
+{
+    SWSS_LOG_ENTER();
+    Port   port;
+    string table_name = consumer.getTableName();
+
+    if (table_name == APP_SFLOW_TABLE_NAME)
+    {
+        sflowStatusSet(consumer);
+        return;
+    }
+    if (!gPortsOrch->allPortsReady())
+    {
+        return;
+    }
+
+    auto it = consumer.m_toSync.begin();
+    while (it != consumer.m_toSync.end())
+    {
+        auto tuple = it->second;
+        string op = kfvOp(tuple);
+        string alias = kfvKey(tuple);
+
+        gPortsOrch->getPort(alias, port);
+        if (op == SET_COMMAND)
+        {
+            bool      admin_state = false;
+            uint32_t  rate       = 0;
+
+            if (!sflowStatus)
+            {
+                return;
+            }
+            auto sflowInfo = m_sflowPortInfoMap.find(port.m_port_id);
+            if (sflowInfo != m_sflowPortInfoMap.end())
+            {
+                rate = sflowSessionGetRate(sflowInfo->second.m_sample_id);
+                admin_state = sflowInfo->second.admin_state;
+            }
+
+            sflowExtractFvs(kfvFieldsValues(tuple), admin_state, rate);
```

Comment:
So it just overwrites the `admin_state` and `rate `retrieved from above. I'm not fully understanding the case here. Do you intend `else `here?

Reply:
If there is an already existing session, then the current admin and rate are initialized from it. Then based on key value pair parsing, we fill the admin_state and rate and override it.
In the logic below, the existing session admin state and rate are compared with the above values. If there is difference, then the session is updated, else it is ignored. 
if (admin_state != sflowInfo->second.admin_state)

if (rate != sflowSessionGetRate(sflowInfo->second.m_sample_id))
For e.g current session has admin=true and rate=100
User configured new rate as 200
The in above logic first the admin is initialized to "true" and rate to 100
in key value parsing, admin is not re-initialized while rate is set to 200
In the checks below, admin state is not changed so no reprogramming is done, while rate is changed it is updated.